src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+    return true;
+}
+
+static bool sendMissingTrustedTimeSource(chip::EndpointId endpointId)
+{
+    Events::MissingTrustedTimeSource::Type event;
+    EventNumber eventNumber;
+
+    CHIP_ERROR error = LogEvent(event, endpointId, eventNumber);
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "Unable to send sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+        return false;
+    }
+
+    // re-schedule event for after min 1hr if TTS is null or cannot be reached
+    ChipLogProgress(Zcl, "Emit sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+    return true;
+}
+
+static void utcTimeChanged(uint64_t utcTime)
+{
+    System::Clock::Seconds32 lastKnownGoodChipEpoch;
+    System::Clock::Microseconds64 realTime;
+    uint32_t utcTimetoChipEpoch;
+
+    Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch);
+    System::SystemClock().GetClock_RealTime(realTime);
+    ChipLogError(Zcl, "UTCTime: %llu Last Known Good Time: %u Real Time: %llu", utcTime, lastKnownGoodChipEpoch.count(),
+                 realTime.count());
+    chip::UnixEpochToChipEpochTime((uint32_t)(utcTime / chip::kMicrosecondsPerSecond), utcTimetoChipEpoch);
+
+    if (utcTimetoChipEpoch >= lastKnownGoodChipEpoch.count()) // update Last Known Good Time if a more recent time is obtained
+    {
+        Server::GetInstance().GetFabricTable().SetLastKnownGoodChipEpochTime(System::Clock::Seconds32(utcTimetoChipEpoch));
+        System::SystemClock().SetClock_RealTime(System::Clock::Microseconds64(utcTime));
+
+        Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch);
+        System::SystemClock().GetClock_RealTime(realTime);
+        ChipLogError(Zcl, " Last Known Good Time: %u Real Time: %llu", lastKnownGoodChipEpoch.count(), realTime.count());
+    }
+}
+
+bool emberAfTimeSynchronizationClusterSetUtcTimeCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetUtcTime::DecodableType & commandData)
+{
+    Optional<StatusCode> status = Optional<StatusCode>::Missing();
+    Status globalStatus         = Status::Success;
+
+    auto utcTime     = commandData.utcTime;
+    auto granularity = commandData.granularity;
+    auto timeSource  = commandData.timeSource;
+
+    TimeSynchronization::GranularityEnum currentGranularity;
+    Granularity::Get(commandPath.mEndpointId, &currentGranularity);
+
+    if (currentGranularity == TimeSynchronization::GranularityEnum::kNoTimeGranularity || granularity >= currentGranularity)
+    {
+        UTCTime::Set(commandPath.mEndpointId, utcTime);
+        utcTimeChanged(utcTime);
+        Granularity::Set(commandPath.mEndpointId, granularity);
+        if (timeSource.HasValue())
+        {
+            TimeSource::Set(commandPath.mEndpointId, timeSource.Value());
```

Comment:
This does not match the spec.

Reply:
Do you mean it should always be set to Admin? In this case, why is it necessary to have this field?

Reply:
Ok, in that case the time source field isn't required as part of the command since it defaults to Admin?

Reply:
The time source in the command is the time source of the command sender. The node receiving the command can use that information to determine if it wants to take the incoming time, or if it already has a time, if it prefers to use its own. So, ex if the receiving node already has a time from NTP and it receive a command that says the source is SNTP and it has some lackluster granularity, it probably doesn't want that and has the option to say essentially "no thanks, I'm good" by sending back a TimeNotAccepted error. 

If the node doesn't have a time, it's required to take the time because something is always better than nothing. Either way, the node sets its OWN source to admin to indicate its using an admin source for time calculation.

Reply:
Granularity works the same way. The incoming granularity tells you the admin's estimate of its granularity. Because SetUTCTime is nothing more than an standard invoke action, we're not accounting for network delay in any way. (ditto if you read from another cluster). So no matter what the incoming granularity is, the granularity that gets set is probably going to be "SecondsGranularity" unless you're expecting large clock drift, in which case you can adjust as you go or just set to "MinutesGranularity" in anticipation. Realistically, most devices won't get that bad even if they're in and out of sleep mode, but it basically means "I'm dodgy and not doing a lot to keep in sync"